src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	for (lod = 4; lod <= 8; lod++)
+	{
+		if (t->lodmask & (1 << (lod - 1)))
+		{
+			UINT32 size = ((t->wmask >> (lod - 1)) + 1) * ((t->hmask >> (lod - 1)) + 1);
+			if (size < 4) size = 4;
+			base += size << bppscale;
+		}
+		t->lodoffset[lod] = base & t->mask;
+	}
+
+	/* set the NCC lookup appropriately */
+	t->texel[1] = t->texel[9] = t->ncc[TEXMODE_NCC_TABLE_SELECT(t->reg[textureMode].u)].texel;
+
+	/* pick the lookup table */
+	t->lookup = t->texel[TEXMODE_FORMAT(t->reg[textureMode].u)];
+
+	/* compute the detail parameters */
+	t->detailmax = TEXDETAIL_DETAIL_MAX(t->reg[tDetail].u);
+	t->detailbias = (INT8)(TEXDETAIL_DETAIL_BIAS(t->reg[tDetail].u) << 2) << 6;
+	t->detailscale = TEXDETAIL_DETAIL_SCALE(t->reg[tDetail].u);
+
+	/* no longer dirty */
+	t->regdirty = false;
+
+	/* check for separate RGBA filtering */
+	assert(!TEXDETAIL_SEPARATE_RGBA_FILTER(t->reg[tDetail].u));
+	//if (TEXDETAIL_SEPARATE_RGBA_FILTER(t->reg[tDetail].u))
+	//	E_Exit("Separate RGBA filters!"); // voodoo 2 feature not implemented
+}
+
+static void prepare_tmu(tmu_state *t)
+{
+	INT64 texdx, texdy;
+	INT32 lodbase;
+
+	/* if the texture parameters are dirty, update them */
+	if (t->regdirty)
+	{
+		recompute_texture_params(t);
+
+		/* ensure that the NCC tables are up to date */
+		if ((TEXMODE_FORMAT(t->reg[textureMode].u) & 7) == 1)
+		{
+			ncc_table *n = &t->ncc[TEXMODE_NCC_TABLE_SELECT(t->reg[textureMode].u)];
+			t->texel[1] = t->texel[9] = n->texel;
+			if (n->dirty)
+				ncc_table_update(n);
+		}
+	}
+
+	/* compute (ds^2 + dt^2) in both X and Y as 28.36 numbers */
+	texdx = (INT64)(t->dsdx >> 14) * (INT64)(t->dsdx >> 14) + (INT64)(t->dtdx >> 14) * (INT64)(t->dtdx >> 14);
+	texdy = (INT64)(t->dsdy >> 14) * (INT64)(t->dsdy >> 14) + (INT64)(t->dtdy >> 14) * (INT64)(t->dtdy >> 14);
+
+	/* pick whichever is larger and shift off some high bits -> 28.20 */
+	if (texdx < texdy)
+		texdx = texdy;
+	texdx >>= 16;
+
+	/* use our fast reciprocal/log on this value; it expects input as a */
+	/* 16.32 number, and returns the log of the reciprocal, so we have to */
+	/* adjust the result: negative to get the log of the original value */
+	/* plus 12 to account for the extra exponent, and divided by 2 to */
+	/* get the log of the square root of texdx */
+	(void)fast_reciplog(texdx, &lodbase);
+	t->lodbasetemp = (-lodbase + (12 << 8)) / 2;
+}
+
+static inline INT32 round_coordinate(float value)
+{
+	return (value > 0 ? (INT32)(value + .4999999f) : (INT32)(value - .5f));
+	//INT32 result = (INT32)floor(value);
+	//return result + (value - (float)result > 0.5f);
+}
+
+/*************************************
+ *
+ *  Statistics management
+ *
+ *************************************/
+
+static void sum_statistics(stats_block *target, const stats_block *source)
+{
+	target->pixels_in += source->pixels_in;
+	target->pixels_out += source->pixels_out;
+	target->chroma_fail += source->chroma_fail;
+	target->zfunc_fail += source->zfunc_fail;
+	target->afunc_fail += source->afunc_fail;
```

Comment:
Good catch! Line 4221 should probably be `stats_block my_stats = {0};` instead of `stats_block my_stats;`.
I was always wondering how important those stats are. Because I'm pretty sure in the OpenGL and Glide wrapper modes of the original patch these stats were not counted at all.
But they are available to the guest system via the registers `fbiPixelsIn`, `fbiChromaFail`, `fbiZfuncFail`, `fbiAfuncFail`, `fbiPixelsOut`, `fbiTrianglesOut`. No idea if any real-world application relies on them or even accesses them. One could add logging on access and then try a whole lot of 3dfx applications (and perhaps Windows9x driver versions).

Reply:
I'd argue it's for profilers or analyzers, not ordinary end-user software (games).